infrastructure/src/main/java/org/corfudb/infrastructure/CompactorService.java

```diff
@@ -35,36 +36,74 @@
 public class CompactorService implements ManagementService {
 
     private final ServerContext serverContext;
-    private final SingletonResource<CorfuRuntime> runtimeSingletonResource;
-    private final ScheduledExecutorService orchestratorThread;
+    private final Duration triggerInterval;
     private final InvokeCheckpointing checkpointerJvmManager;
     private final CompactionTriggerPolicy compactionTriggerPolicy;
-
+    private final ScheduledExecutorService orchestratorThread;
+    private final TrimLog trimLog;
     private Optional<CompactorLeaderServices> optionalCompactorLeaderServices = Optional.empty();
     private Optional<CorfuStore> optionalCorfuStore = Optional.empty();
     private Optional<DistributedCheckpointerHelper> optionalDistributedCheckpointerHelper = Optional.empty();
-    private TrimLog trimLog;
+    private Optional<CorfuRuntime> corfuRuntimeOptional = Optional.empty();
+    private ScheduledFuture<?> scheduledFuture;
     private final Logger log;
     private static final Duration LIVENESS_TIMEOUT = Duration.ofMinutes(1);
+    private static final int SYSTEM_DOWN_HANDLER_TRIGGER_LIMIT = 60;
 
-    CompactorService(@NonNull ServerContext serverContext,
-                     @NonNull SingletonResource<CorfuRuntime> runtimeSingletonResource,
+    public CompactorService(@NonNull ServerContext serverContext,
+                     @NonNull Duration triggerInterval,
                      @NonNull InvokeCheckpointing checkpointerJvmManager,
                      @NonNull CompactionTriggerPolicy compactionTriggerPolicy) {
         this.serverContext = serverContext;
-        this.runtimeSingletonResource = runtimeSingletonResource;
-
+        this.triggerInterval = triggerInterval;
+        this.checkpointerJvmManager = checkpointerJvmManager;
+        this.compactionTriggerPolicy = compactionTriggerPolicy;
         this.orchestratorThread = Executors.newSingleThreadScheduledExecutor(
                 new ThreadFactoryBuilder()
                         .setNameFormat("Cmpt-" + serverContext.getServerConfig().get("<port>") + "-chkpter")
                         .build());
-        this.checkpointerJvmManager = checkpointerJvmManager;
-        this.compactionTriggerPolicy = compactionTriggerPolicy;
+        this.trimLog = new TrimLog();
         this.log = LoggerFactory.getLogger("compactor-leader");
     }
 
-    private CorfuRuntime getCorfuRuntime() {
-        return runtimeSingletonResource.get();
+    private Runnable getSystemDownHandlerForCompactor() {
+        return () -> {
+            synchronized (this) {
+                log.warn("CorfuRuntime for CompactorService stalled. Invoking systemDownHandler after {} "
+                        + "unsuccessful tries.", SYSTEM_DOWN_HANDLER_TRIGGER_LIMIT);
+                shutdown();
+                start(this.triggerInterval);
+            }
```

Comment:
This will serialize the re-initialization, but doesn't guarantee one-time initialization. Essentially, if three threads encounter a `UnreachableClusterException` for the same client, then all three threads will invoke the re-initialization logic in serial order creating three different runtimes. As a side-effect multiple calls to `getNewCorfuRuntime()` can observe different versions of the client, but instead we should coalescence the calls and only reinitialize once.

To coalesce the handler calls you can keep the reference of the current runtime, once the handler is called you re-initialize the client only if the current reference != the one being signaled from the handler, that way only the first call will create a new client and the subsequent handler calls become no-ops.

Reply:
The shutdown will anyway be killing the runtimes created back to back, but I see your point to reinitialize only once... But the condition to reinitialize should be when current runtime is same as the runtime from the handler, and not unequal right?